src/runtime/virtcontainers/qemu_arch_base.go

```diff
@@ -318,29 +319,49 @@ func (q *qemuArchBase) memoryTopology(memoryMb, hostMemoryMb uint64, slots uint8
 }
 
 func (q *qemuArchBase) appendConsole(_ context.Context, devices []govmmQemu.Device, path string) ([]govmmQemu.Device, error) {
-	serial := govmmQemu.SerialDevice{
-		Driver:        govmmQemu.VirtioSerial,
-		ID:            "serial0",
-		DisableModern: q.nestedRun,
-		MaxPorts:      uint(2),
-	}
+	var serial, console govmmQemu.Device
+	var consoleKernelParams []Param
 
-	devices = append(devices, serial)
+	if q.legacySerial {
+		serial = govmmQemu.LegacySerialDevice{
+			Chardev: "charconsole0",
+		}
 
-	console := govmmQemu.CharDevice{
-		Driver:   govmmQemu.Console,
-		Backend:  govmmQemu.Socket,
-		DeviceID: "console0",
-		ID:       "charconsole0",
-		Path:     path,
-	}
+		console = govmmQemu.CharDevice{
+			Driver:   govmmQemu.LegacySerial,
+			Backend:  govmmQemu.Socket,
+			DeviceID: "console0",
+			ID:       "charconsole0",
+			Path:     path,
+		}
 
-	devices = append(devices, console)
+		consoleKernelParams = []Param{
+			{"console", "ttyS0"},
+		}
+	} else {
+		serial = govmmQemu.SerialDevice{
+			Driver:        govmmQemu.VirtioSerial,
+			ID:            "serial0",
+			DisableModern: q.nestedRun,
+			MaxPorts:      uint(2),
+		}
+
+		console = govmmQemu.CharDevice{
+			Driver:   govmmQemu.Console,
+			Backend:  govmmQemu.Socket,
+			DeviceID: "console0",
+			ID:       "charconsole0",
+			Path:     path,
+		}
```

Comment:
Except for the `Driver` field, `console` has the same content as in the legacy serial case. Maybe factor some more ?

Reply:
Indeed possible, i wonder if it would be better to leave it as just for the sake of readability